File: app/src/main/java/com/github/goldy1992/mp3player/service/library/ContentSearchers.java
package com.github.goldy1992.mp3player.service.library;

import com.github.goldy1992.mp3player.commons.MediaItemType;
import com.github.goldy1992.mp3player.service.library.content.searcher.ContentSearcher;
import com.github.goldy1992.mp3player.service.library.content.searcher.FolderSearcher;
import com.github.goldy1992.mp3player.service.library.content.searcher.SongSearcher;

import java.util.Collection;
import java.util.EnumMap;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import static com.github.goldy1992.mp3player.commons.MediaItemType.FOLDER;

@Singleton
public class ContentSearchers {

    /** */
    EnumMap<MediaItemType, ContentSearcher> contentSearcherMap;

    @Inject
    public ContentSearchers(SongSearcher songSearcher,
                            FolderSearcher folderSearcher) {
        EnumMap<MediaItemType, ContentSearcher> map = new EnumMap<>(MediaItemType.class);
        map.put(MediaItemType.SONG, songSearcher);
        map.put(FOLDER, folderSearcher);
        this.contentSearcherMap = map;
    }

    public ContentSearcher get(MediaItemType mediaItemType) {
        return this.contentSearcherMap.get(mediaItemType);
    }

    public Collection<ContentSearcher> getAll() {
        return contentSearcherMap.values();
    }

}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/content/searcher/ContentSearchers.kt
package com.github.goldy1992.mp3player.service.library.content.searcher

import com.github.goldy1992.mp3player.commons.MediaItemType
import dagger.hilt.android.scopes.ServiceScoped
import java.util.EnumMap
import javax.inject.Inject

@ServiceScoped
class ContentSearchers

    @Inject
    constructor(songSearcher: SongSearcher,
               folderSearcher: FolderSearcher,
                albumSearcher: AlbumSearcher) {
    /**  */
    var contentSearcherMap: EnumMap<MediaItemType, ContentSearcher>

    operator fun get(mediaItemType: MediaItemType?): ContentSearcher? {
        return contentSearcherMap[mediaItemType]
    }

    val all: Collection<ContentSearcher>
        get() = contentSearcherMap.values

    init {
        val map = EnumMap<MediaItemType, ContentSearcher>(MediaItemType::class.java)
        map[MediaItemType.SONG] = songSearcher
        map[MediaItemType.FOLDER] = folderSearcher
        map[MediaItemType.ALBUMS] = albumSearcher
        contentSearcherMap = map
    }
}